drivers/ad9208/ad9208.c

```diff
@@ -0,0 +1,514 @@
+	if (!dev)
+		return -ENOMEM;
+
+ 	dev->st = (struct ad9208_state *)calloc(1, sizeof(*dev->st));
+	if (!dev->st) {
+		ret = -ENOMEM;
+		goto error_1;
+	}
+ 	/* SPI */
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+ 	if (ret < 0)
+ 		goto error_2;
+
+	/* GPIO */
+	ret = gpio_get(&dev->gpio_powerdown, init_param->gpio_powerdown);
+	if (ret < 0)
+		goto error_3;
+
+	ret = gpio_direction_output(dev->gpio_powerdown, GPIO_LOW);
+	if (ret < 0)
+		goto error_4;
+
+	ret = gpio_set_value(dev->gpio_powerdown, GPIO_LOW);
+	if (ret < 0)
+		goto error_4;
+
+ 	struct ad9208_state *st = (struct ad9208_state *)calloc(1, sizeof(*st));
+	if (!st) {
+		ret = -ENOMEM;
+		goto error_4;
+	}
+
+	st->jesd_param = (jesd_param_t *)calloc(1, sizeof(*st->jesd_param));
+	if (!st->jesd_param) {
+		ret = -ENOMEM;
+		goto error_5;
+	}
+
+	st->adc_h = (ad9208_handle_t *)calloc(1, sizeof(*st->adc_h));
+	if (!st->adc_h) {
+		ret = -ENOMEM;
+		goto error_6;
+	}
+
+	st->adc_h->user_data = dev->spi_desc;
+	st->adc_h->dev_xfer = ad9208_spi_xfer;
+	st->adc_h->delay_us = ad9208_udelay;
+
+	st->sampling_frequency_hz = init_param->sampling_frequency_hz;
+	st->input_div = init_param->input_div;
+	st->duty_cycle_stabilizer_en = init_param->duty_cycle_stabilizer_en;
+	st->powerdown_mode = init_param->powerdown_mode;
+	st->powerdown_pin_en = init_param->powerdown_pin_en;
+	st->current_scale = init_param->current_scale;
+	st->analog_input_mode = init_param->analog_input_mode;
+	st->ext_vref_en = init_param->ext_vref_en;
+	st->buff_curr_n = init_param->buff_curr_n;
+	st->buff_curr_p = init_param->buff_curr_p;
+	st->fc_ch = init_param->fc_ch;
+	st->ddc_cnt = init_param->ddc_cnt;
+	for (i = 0; i < st->ddc_cnt; i++) {
+		st->ddc[i].decimation = init_param->ddc[i].decimation;
+		st->ddc[i].nco_mode = init_param->ddc[i].nco_mode;
+		st->ddc[i].carrier_freq_hz = init_param->ddc[i].carrier_freq_hz;
+		st->ddc[i].gain_db = init_param->ddc[i].gain_db;
+		st->ddc[i].po = init_param->ddc[i].po;
+	}
+	st->ddc_output_format_real_en = init_param->ddc_output_format_real_en;
+	st->ddc_input_format_real_en = init_param->ddc_input_format_real_en;
+	st->test_mode_ch0 = init_param->test_mode_ch0;
+	st->test_mode_ch1 = init_param->test_mode_ch1;
+
+	/* SYSREF Config */
+	st->sysref_lmfc_offset = init_param->sysref_lmfc_offset;
+	st->sysref_clk_edge_sel = init_param->sysref_clk_edge_sel;
+	st->sysref_edge_sel = init_param->sysref_edge_sel;
+	st->sysref_neg_window_skew = init_param->sysref_neg_window_skew;
+	st->sysref_pos_window_skew = init_param->sysref_pos_window_skew;
+	st->sysref_mode = init_param->sysref_mode;
+	st->sysref_count = init_param->sysref_count;
+
+	st->jesd_param = init_param->jesd_param;
+	st->jesd_subclass = init_param->jesd_subclass;
+
+	ret = ad9208_setup(st);
+	if (ret < 0) {
+		printf("Failed to setup device\n");
+		goto error_7;
+	}
+
+ 	dev->st = st;
+
+ 	printf("AD9208 successfully initialized\n");
+
+	*device = dev;
+
+ 	return 0;
+
+error_7:
```

Comment:
as a coding-pattern, these `goto error_` groups look hard to maintain/change;
that is: they require a bit of brain power to follow and make sure they are not leaking;
for code that is final, and is sure to not change, it's probably fine; their speed/run-time efficiency remains uncontested;

not sure how often this code would change to make things sure we don't introduce leaks/errors at a later change;

the alternative here, is to create a single `error` label and do:
```
    if (st->adc_h)
        free(st->adc_h);
    if (st->jesd_param)
        free(st->jesd_param);
    ....
```

of course, you'd need to make sure that `st` is initialized with zeros, so that non-NULL is free'd properly;
and the null-checks are required here, as not all implementations of `free()` are guaranteed to ignore `free(NULL)`